pylot/control/mpc/mpc_agent_operator.py

```diff
@@ -0,0 +1,333 @@
+                    new_speed_factor_p = pylot.control.utils.stop_pedestrian(
+                        ego_vehicle_location,
+                        pedestrian.transform.location,
+                        wp_vector,
+                        speed_factor_p,
+                        self._flags)
+                    speed_factor_p = min(speed_factor_p, new_speed_factor_p)
+
+        if self._flags.stop_for_traffic_lights:
+            for tl in traffic_lights:
+                if (self._map.must_obbey_traffic_light(
+                        ego_vehicle_location, tl.transform.location) and
+                        self._is_traffic_light_visible(
+                            ego_vehicle_location, tl.transform.location)):
+                    new_speed_factor_tl = pylot.control.utils.stop_traffic_light(
+                        ego_vehicle_location,
+                        tl.transform.location,
+                        tl.state,
+                        wp_vector,
+                        wp_angle,
+                        speed_factor_tl,
+                        self._flags)
+                    speed_factor_tl = min(speed_factor_tl, new_speed_factor_tl)
+
+        speed_factor = min(speed_factor_tl, speed_factor_p, speed_factor_v)
+
+        # slow down around corners
+        if math.fabs(wp_angle_speed) < 0.1:
+            speed_factor = 0.3 * speed_factor
+
+        state = {
+            'stop_pedestrian': speed_factor_p,
+            'stop_vehicle': speed_factor_v,
+            'stop_traffic_lights': speed_factor_tl
+        }
+
+        return speed_factor, state
+
+    def setup_mpc(self, waypoints):
+        path = np.array([[wp.location.x, wp.location.y] for wp in waypoints])
+
+        # convert target waypoints into spline
+        spline = CubicSpline2D(path[:, 0], path[:, 1], self._logger)
+        ss = []
+        vels = []
+        xs = []
+        ys = []
+        yaws = []
+        ks = []
+        for s in spline.s[:-2]:
+            x, y = spline.calc_position(s)
+            yaw = np.abs(spline.calc_yaw(s))
+            k = spline.calc_curvature(s)
+            xs.append(x)
+            ys.append(y)
+            yaws.append(yaw)
+            ks.append(k)
+            ss.append(s)
+            vels.append(18)
+
+        self._config["reference"] = {
+            't_list': [],  # Time [s]
+            's_list': ss,  # Arc distance [m]
+            'x_list': xs,  # Desired X coordinates [m]
+            'y_list': ys,  # Desired Y coordinates [m]
+            'k_list': ks,  # Curvatures [1/m]
+            'vel_list': vels,  # Desired tangential velocities [m/s]
+            'yaw_list': yaws,  # Yaws [rad]
+        }
+        # draw intended trajectory
+        for p in path:
+            self._world.debug.draw_point(carla.Location(x=p[0], y=p[1], z=0.5),
+                                         size=0.2,
+                                         life_time=1.0, color=carla.Color(0, 0, 255))
```

Comment:
This should be removed. We can have a visualization operator later that shows the trajectory.

Reply:
Will defer to a later PR!